### ISSUE TITLE
string(0z01) returns [0x01] but should return 0z01

### ISSUE THREAD

**Issue by dpelle** [vim/vim#3807]
`string(0z01)` returns `"[0x01]"` but it should I think return `"0z01"`.

`:help string()`, says: 
```
string({expr})	Return {expr} converted to a String.  If {expr} is a Number,
		Float, String or a composition of them, then the result can be
		parsed back with |eval()|.
```

I see that `:echo eval(string(0z01))` prints `[1]` which is a list and not a blob.
If instead `string(0z01)` returned `"0x01"`, then  `eval(string(0z01))` would properly
parse back into a blob.

**Comment by mattn** [comment]
Agreed. However hex string is useful to manipulate binary dump. So how about this patch?

```diff
diff --git a/src/blob.c b/src/blob.c
index 6b771fed1..d2b6d89c8 100644
--- a/src/blob.c
+++ b/src/blob.c
@@ -179,17 +179,15 @@ blob2string(blob_T *blob, char_u **tofree, char_u *numbuf)
     if (blob == NULL)
     {
 	*tofree = NULL;
-	return (char_u *)"[]";
+	return (char_u *)"0z";
     }
 
     // Store bytes in the growarray.
     ga_init2(&ga, 1, 4000);
-    ga_append(&ga, '[');
+    ga_concat(&ga, (char_u*)"0z");
     for (i = 0; i < blob_len(blob); i++)
     {
-	if (i > 0)
-	    ga_concat(&ga, (char_u *)",");
-	vim_snprintf((char *)numbuf, NUMBUFLEN, "0x%02X", (int)blob_get(blob, i));
+	vim_snprintf((char *)numbuf, NUMBUFLEN, "%02X", (int)blob_get(blob, i));
 	ga_concat(&ga, numbuf);
     }
     ga_append(&ga, ']');
diff --git a/src/json.c b/src/json.c
index 11bb04832..09d3fcfcf 100644
--- a/src/json.c
+++ b/src/json.c
@@ -246,7 +246,7 @@ json_encode_item(garray_T *gap, typval_T *val, int copyID, int options)
 		{
 		    if (i > 0)
 			ga_concat(gap, (char_u *)",");
-		    vim_snprintf((char *)numbuf, NUMBUFLEN, "%d",
+		    vim_snprintf((char *)numbuf, NUMBUFLEN, "0x%02X",
 			    (int)blob_get(b, i));
 		    ga_concat(gap, numbuf);
 		}
diff --git a/src/testdir/test_blob.vim b/src/testdir/test_blob.vim
index ec3de89d1..2a04d2234 100644
--- a/src/testdir/test_blob.vim
+++ b/src/testdir/test_blob.vim
@@ -81,9 +81,9 @@ endfunc
 
 func Test_blob_to_string()
   let b = 0zDEADBEEF
-  call assert_equal('[0xDE,0xAD,0xBE,0xEF]', string(b))
+  call assert_equal('0zDEADBEEF]', string(b))
   call remove(b, 0, 3)
-  call assert_equal('[]', string(b))
+  call assert_equal('0z', string(b))
 endfunc
 
 func Test_blob_compare()
@@ -253,7 +253,7 @@ func Test_blob_reverse()
 endfunc
 
 func Test_blob_json_encode()
-  call assert_equal('[222,173,190,239]', json_encode(0zDEADBEEF))
+  call assert_equal('[0xDE,0xED,0xBE,0xEF]', json_encode(0zDEADBEEF))
   call assert_equal('[]', json_encode(0z))
 endfunc
 
```

**Comment by dpelle** [comment]
@k-takata wrote:

> Agreed. However hex string is useful to manipulate binary dump.
> So how about this patch?

Hmmm, json should have hex numbers.
So `json_encode(0zDEADBEEF)` should rather give `'[222,173,190,239]'` and not `'[0xDE,0xED,0xBE,0xEF]'`.

**Comment by brammool** [comment]
> Agreed. However hex string is useful to manipulate binary dump. So how
> about this patch?

I had also been wondering about the string form of a blob.  It would be
good if it can be parsed back.  Thus using the 0z form is useful.
The form we use so far looks too much like a list.

However, the 0z form is quite unreadable.  How about using dots
to make it readable:
	0z00112233.44556677.8899aabb

Groups of 32 bits are probably best for readability.
When parsing, we can allow for dots anywhere, so the user could also do:
	0z0011.2233.4455.6677.8899.aabb
Or even:
	0z00.11.22.33.44.55.66.77.88.99.aa.bb

> --- a/src/json.c
> +++ b/src/json.c
> @@ -246,7 +246,7 @@ json_encode_item(garray_T *gap, typval_T *val, int copyID, int options)
>  		{
>  		    if (i > 0)
>  			ga_concat(gap, (char_u *)",");
> -		    vim_snprintf((char *)numbuf, NUMBUFLEN, "%d",
> +		    vim_snprintf((char *)numbuf, NUMBUFLEN, "0x%02X",

I don't think json supports 0x12 for numbers.


-- 
hundred-and-one symptoms of being an internet addict:
230. You spend your Friday nights typing away at your keyboard

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by dpelle** [comment]
Fixed in 4131fd5509b283e978e8c6161f09643b64719787
